### ISSUE TITLE
Pretraining OOM error

### ISSUE THREAD

**Issue by Optimox** [dreamquark-ai/tabnet#341]
**What is the current behavior?**

I've noticed that you can get OOM error on GPU depending on the size of the validation set when performing pretraining.

This does not make sense, need to investigate. I will add more details later.

**Comment by eduardocarvp** [comment]
To give more details on this:

I managed to replicate the error on GPU, by using a large validation set. The problem consists on stacking all the answers from valid batches into a single `torch` tensor. On `predict_epoch` we do:

https://github.com/dreamquark-ai/tabnet/blob/5ac55834b32693abc4b22028a74475ee0440c2a5/pytorch_tabnet/pretraining.py#L342-L350

and `stack_batches`:

https://github.com/dreamquark-ai/tabnet/blob/5ac55834b32693abc4b22028a74475ee0440c2a5/pytorch_tabnet/pretraining.py#L374-L378

So in the end we have three large torch tensors that end up using all the GPU memory.

If we want to keep the door open to any evaluation metric, one solution would be to convert these values to numpy arrays before concatenation and reimplement the unsupervised loss using numpy operations instead of torch. It would still insert a high load on RAM, but I guess it's better than overloading the GPU. We can probably find a way to store this information in a more effecient way as well, even if we need some additional operations to compute the evaluation metric.

**Comment by Optimox** [comment]
We should also change the problem of dividing by the standard deviation + epsilon.
It would probably be much more stable to divide by something like standard deviation +1 to avoid the loss to skyrocket.
